File: leetcode2300/leetcode2300.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
    public:
        vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
            sort(potions.begin(), potions.end());
            for (int &x : spells){
                long long target = (success - 1) / x;
                if (target < potions.back()){
                    x = potions.end() - upper_bound(potions.begin(), potions.end(), (int)target);
                } else {
                    x = 0;
                }
            }
            return spells;
        }
    };

File: leetcode2300/leetcode2300.py
from bisect import bisect_right


class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        m = len(potions)
        success -= 1
        return [m - bisect_right(potions, success // x) for x in spells]

